File: backend/app/schemas.py
# Mirrors @lightsale/shared Zod schemas — keep field names in sync with packages/shared/src/schemas.ts

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from app.document_normalize import default_output_settings, normalize_project_document

VALID_ROOM_TYPES = frozenset(
    {
        "living_room",
        "kitchen",
        "dining_room",
        "bedroom",
        "bathroom",
        "hallway",
        "home_office",
        "open_office",
        "private_office",
        "meeting_room",
        "reception",
        "corridor",
        "storage",
        "toilet",
        "technical_room",
        "other",
    }
)

VALID_STYLE_PRESETS = frozenset(
    {
        "functional",
        "warm_modern",
        "minimal",
        "hotel_chic",
        "industrial",
        "architectural",
        "custom",
    }
)

LEGACY_ROOM_TYPE_MAP = {
    "warehouse": "storage",
    "office": "open_office",
    "production": "technical_room",
    "corridor": "corridor",
}

LEGACY_STYLE_PRESET_MAP = {
    "standard": "functional",
    "high-bay": "industrial",
    "ambient": "warm_modern",
    "task-focused": "functional",
    "custom": "custom",
}

RoomType = Literal[
    "living_room",
    "kitchen",
    "dining_room",
    "bedroom",
    "bathroom",
    "hallway",
    "home_office",
    "open_office",
    "private_office",
    "meeting_room",
    "reception",
    "corridor",
    "storage",
    "toilet",
    "technical_room",
    "other",
]
CeilingType = Literal["suspended", "exposed", "grid", "sloped", "other"]
StylePreset = Literal[
    "functional",
    "warm_modern",
    "minimal",
    "hotel_chic",
    "industrial",
    "architectural",
    "custom",
]


class PointModel(BaseModel):
    x: float
    y: float


class ScaleCalibrationModel(BaseModel):
    pointA: PointModel
    pointB: PointModel
    realDistanceMetres: float = Field(gt=0)


class RoomModel(BaseModel):
    id: UUID
    name: str = Field(min_length=1)
    vertices: list[PointModel] = Field(min_length=3)
    roomType: RoomType = "other"
    ceilingHeightMetres: float = Field(default=3.0, gt=0)
    ceilingType: CeilingType = "exposed"
    targetLux: float | None = None
    stylePreset: StylePreset = "functional"
    selectedProductId: str | None = None
    utilisationFactor: float = Field(default=0.6, gt=0, le=1)
    maintenanceFactor: float = Field(default=0.8, gt=0, le=1)

    @model_validator(mode="before")
    @classmethod
    def migrate_legacy_fields(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        next_data = dict(data)
        room_type = next_data.get("roomType")
        if isinstance(room_type, str) and room_type not in VALID_ROOM_TYPES:
            next_data["roomType"] = LEGACY_ROOM_TYPE_MAP.get(room_type, "other")
        style = next_data.get("stylePreset")
        if isinstance(style, str) and style not in VALID_STYLE_PRESETS:
            next_data["stylePreset"] = LEGACY_STYLE_PRESET_MAP.get(
                style, "functional"
            )
        if not isinstance(next_data.get("utilisationFactor"), (int, float)):
            next_data["utilisationFactor"] = 0.6
        if not isinstance(next_data.get("maintenanceFactor"), (int, float)):
            next_data["maintenanceFactor"] = 0.8
        product_id = next_data.get("selectedProductId")
        if product_id is not None and not isinstance(product_id, str):
            next_data["selectedProductId"] = None
        return next_data

    @field_validator("targetLux")
    @classmethod
    def validate_target_lux(cls, value: float | None) -> float | None:
        if value is not None and value <= 0:
            raise ValueError("targetLux must be positive when set")
        return value


class ViewportStateModel(BaseModel):
    zoom: float = Field(gt=0)
    offsetX: float
    offsetY: float


PlacementSource = Literal["generated", "manual"]


class LuminaireModel(BaseModel):
    id: UUID
    roomId: UUID
    productId: str = Field(min_length=1)
    x: float
    y: float
    rotationDegrees: float
    placementSource: PlacementSource
    createdAt: datetime


class FloorPlanAssetModel(BaseModel):
    id: UUID
    fileName: str
    mimeType: str
    widthPx: float = Field(gt=0)
    heightPx: float = Field(gt=0)
    storagePath: str


class OutputSettingsModel(BaseModel):
    projectName: str = Field(default="", max_length=200)
    customerName: str = Field(default="", max_length=200)
    projectReference: str = Field(default="", max_length=200)
    projectAddress: str = Field(default="", max_length=500)
    designerName: str = Field(default="", max_length=200)
    outputDate: str = Field(default="", max_length=50)
    notes: str = Field(default="", max_length=2000)
    showFloorPlanBackground: bool = True
    showRoomOutlines: bool = True
    showRoomNames: bool = True
    showLuminaireSymbols: bool = True
    showLuminaireNumbers: bool = False
    showScale: bool = True
    showLegend: bool = True
    showLightIndicator: bool = False
    includeLightIndicatorInPdf: bool = False
    showLuxSummary: bool = True
    showComplianceStatus: bool = True


class ProjectDocumentModel(BaseModel):
    scale: ScaleCalibrationModel | None = None
    rooms: list[RoomModel] = Field(default_factory=list)
    luminaires: list[LuminaireModel] = Field(default_factory=list)
    outputSettings: OutputSettingsModel = Field(
        default_factory=OutputSettingsModel
    )
    viewport: ViewportStateModel | None = None

    @model_validator(mode="before")
    @classmethod
    def normalize_document(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return normalize_project_document(data)
        return data


class ProjectModel(BaseModel):
    id: UUID
    name: str
    createdAt: datetime
    updatedAt: datetime
    floorPlan: FloorPlanAssetModel | None
    document: ProjectDocumentModel


class CreateProjectRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class UpdateProjectDocumentRequest(BaseModel):
    scale: ScaleCalibrationModel | None = None
    rooms: list[RoomModel] = Field(default_factory=list)
    luminaires: list[LuminaireModel] = Field(default_factory=list)
    outputSettings: OutputSettingsModel = Field(
        default_factory=OutputSettingsModel
    )
    viewport: ViewportStateModel | None = None

    @model_validator(mode="before")
    @classmethod
    def normalize_document(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return normalize_project_document(data)
        return data


class ProjectListItem(BaseModel):
    id: UUID
    name: str
    createdAt: datetime
    updatedAt: datetime
    hasFloorPlan: bool
    customerName: str = ""
    roomCount: int = 0
    luminaireCount: int = 0


class RenameProjectRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class ProjectTrashStats(BaseModel):
    activeProjectCount: int
    trashProjectCount: int
    totalUploadedBytes: int | None = None


def default_document() -> dict[str, Any]:
    return {
        "scale": None,
        "rooms": [],
        "luminaires": [],
        "outputSettings": default_output_settings(),
        "viewport": {"zoom": 1, "offsetX": 0, "offsetY": 0},
    }

File: packages/shared/src/schemas.ts
import { z } from "zod";
import { DEMO_PRODUCT_IDS, getProductById } from "./product-catalog.js";
import { migrateLegacyRoomFields } from "./room-migration.js";

export const PointSchema = z.object({
  x: z.number(),
  y: z.number(),
});

export type Point = z.infer<typeof PointSchema>;

export const ScaleCalibrationSchema = z.object({
  pointA: PointSchema,
  pointB: PointSchema,
  realDistanceMetres: z.number().positive(),
});

export type ScaleCalibration = z.infer<typeof ScaleCalibrationSchema>;

export const ROOM_TYPES = [
  "living_room",
  "kitchen",
  "dining_room",
  "bedroom",
  "bathroom",
  "hallway",
  "home_office",
  "open_office",
  "private_office",
  "meeting_room",
  "reception",
  "corridor",
  "storage",
  "toilet",
  "technical_room",
  "other",
] as const;

export const RoomTypeSchema = z.enum(ROOM_TYPES);
export type RoomType = z.infer<typeof RoomTypeSchema>;

export const CEILING_TYPES = [
  "suspended",
  "exposed",
  "grid",
  "sloped",
  "other",
] as const;

export const CeilingTypeSchema = z.enum(CEILING_TYPES);
export type CeilingType = z.infer<typeof CeilingTypeSchema>;

export const STYLE_PRESETS = [
  "functional",
  "warm_modern",
  "minimal",
  "hotel_chic",
  "industrial",
  "architectural",
  "custom",
] as const;

export const StylePresetSchema = z.enum(STYLE_PRESETS);
export type StylePreset = z.infer<typeof StylePresetSchema>;

export const DEFAULT_ROOM_PROPERTY_VALUES = {
  roomType: "other",
  ceilingHeightMetres: 3,
  ceilingType: "exposed",
  targetLux: null,
  stylePreset: "functional",
  selectedProductId: null,
  utilisationFactor: 0.6,
  maintenanceFactor: 0.8,
} as const satisfies {
  roomType: RoomType;
  ceilingHeightMetres: number;
  ceilingType: CeilingType;
  targetLux: number | null;
  stylePreset: StylePreset;
  selectedProductId: string | null;
  utilisationFactor: number;
  maintenanceFactor: number;
};

export const RoomSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  vertices: z.array(PointSchema).min(3),
  roomType: RoomTypeSchema.default(DEFAULT_ROOM_PROPERTY_VALUES.roomType),
  ceilingHeightMetres: z
    .number()
    .positive()
    .default(DEFAULT_ROOM_PROPERTY_VALUES.ceilingHeightMetres),
  ceilingType: CeilingTypeSchema.default(DEFAULT_ROOM_PROPERTY_VALUES.ceilingType),
  targetLux: z
    .number()
    .positive()
    .nullable()
    .default(DEFAULT_ROOM_PROPERTY_VALUES.targetLux),
  stylePreset: StylePresetSchema.default(DEFAULT_ROOM_PROPERTY_VALUES.stylePreset),
  selectedProductId: z
    .string()
    .min(1)
    .nullable()
    .default(DEFAULT_ROOM_PROPERTY_VALUES.selectedProductId),
  utilisationFactor: z
    .number()
    .positive()
    .max(1)
    .default(DEFAULT_ROOM_PROPERTY_VALUES.utilisationFactor),
  maintenanceFactor: z
    .number()
    .positive()
    .max(1)
    .default(DEFAULT_ROOM_PROPERTY_VALUES.maintenanceFactor),
});

export type Room = z.infer<typeof RoomSchema>;

export type RoomPropertyPatch = Partial<
  Pick<
    Room,
    | "name"
    | "roomType"
    | "ceilingHeightMetres"
    | "ceilingType"
    | "targetLux"
    | "stylePreset"
    | "selectedProductId"
    | "utilisationFactor"
    | "maintenanceFactor"
  >
>;

function sanitizeRoomProductSelection(room: Room): Room {
  if (room.selectedProductId === null) {
    return room;
  }
  if (!DEMO_PRODUCT_IDS.has(room.selectedProductId)) {
    return { ...room, selectedProductId: null };
  }
  const product = getProductById(room.selectedProductId);
  if (product === undefined) {
    return { ...room, selectedProductId: null };
  }
  const compatible =
    product.suitableRoomTypes.includes(room.roomType) &&
    product.suitableStylePresets.includes(room.stylePreset);
  if (!compatible) {
    return { ...room, selectedProductId: null };
  }
  return room;
}

export function normalizeRoom(room: unknown): Room {
  if (typeof room !== "object" || room === null) {
    return RoomSchema.parse(room);
  }
  const migrated = migrateLegacyRoomFields(room as Record<string, unknown>);
  const parsed = RoomSchema.parse(migrated);
  return sanitizeRoomProductSelection(parsed);
}

export function defaultRoomPropertyFields(): Pick<
  Room,
  | "roomType"
  | "ceilingHeightMetres"
  | "ceilingType"
  | "targetLux"
  | "stylePreset"
  | "selectedProductId"
  | "utilisationFactor"
  | "maintenanceFactor"
> {
  return {
    roomType: DEFAULT_ROOM_PROPERTY_VALUES.roomType,
    ceilingHeightMetres: DEFAULT_ROOM_PROPERTY_VALUES.ceilingHeightMetres,
    ceilingType: DEFAULT_ROOM_PROPERTY_VALUES.ceilingType,
    targetLux: DEFAULT_ROOM_PROPERTY_VALUES.targetLux,
    stylePreset: DEFAULT_ROOM_PROPERTY_VALUES.stylePreset,
    selectedProductId: DEFAULT_ROOM_PROPERTY_VALUES.selectedProductId,
    utilisationFactor: DEFAULT_ROOM_PROPERTY_VALUES.utilisationFactor,
    maintenanceFactor: DEFAULT_ROOM_PROPERTY_VALUES.maintenanceFactor,
  };
}

export const ViewportStateSchema = z.object({
  zoom: z.number().positive(),
  offsetX: z.number(),
  offsetY: z.number(),
});

export type ViewportState = z.infer<typeof ViewportStateSchema>;

export const FloorPlanAssetSchema = z.object({
  id: z.string().uuid(),
  fileName: z.string(),
  mimeType: z.string(),
  widthPx: z.number().positive(),
  heightPx: z.number().positive(),
  storagePath: z.string(),
});

export type FloorPlanAsset = z.infer<typeof FloorPlanAssetSchema>;

export const ProjectDocumentSchema = z.object({
  scale: ScaleCalibrationSchema.nullable(),
  rooms: z.array(RoomSchema),
  viewport: ViewportStateSchema.optional(),
});

export type ProjectDocument = z.infer<typeof ProjectDocumentSchema>;

export const ProjectSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  floorPlan: FloorPlanAssetSchema.nullable(),
  document: ProjectDocumentSchema,
});

export type Project = z.infer<typeof ProjectSchema>;

export const CreateProjectInputSchema = z.object({
  name: z.string().min(1).max(200),
});

export type CreateProjectInput = z.infer<typeof CreateProjectInputSchema>;

export const UpdateProjectDocumentInputSchema = z.object({
  scale: ScaleCalibrationSchema.nullable(),
  rooms: z.array(RoomSchema),
  viewport: ViewportStateSchema.optional(),
});

export type UpdateProjectDocumentInput = z.infer<
  typeof UpdateProjectDocumentInputSchema
>;

export const DEFAULT_VIEWPORT: ViewportState = {
  zoom: 1,
  offsetX: 0,
  offsetY: 0,
};

export const EMPTY_PROJECT_DOCUMENT: ProjectDocument = {
  scale: null,
  rooms: [],
  viewport: DEFAULT_VIEWPORT,
};
